lib/src/painters/recorder_wave_painter.dart

```diff
@@ -137,91 +137,86 @@ class RecorderWavePainter extends CustomPainter {
   bool shouldRepaint(RecorderWavePainter oldDelegate) => true;
 
   void _drawTextInRange(Canvas canvas, int i, Size size) {
-    if (_labels.isNotEmpty) {
+    if (_labels.isNotEmpty && i < _labels.length) {
+      final label = _labels[i];
+      final content = label.content;
+      final offset = label.offset;
+      final halfWidth = size.width / 2;
       final textSpan = TextSpan(
-        text: _labels[i].content,
+        text: content,
         style: durationStyle,
       );
 
-      if (_labels[i].offset.dx > -size.width / 2 &&
-          _labels[i].offset.dx < size.width + size.width / 2) {
+      // Text painting is performance intensive process so we will only render
+      // labels whose position is greater then -halfWidth and triple of
+      // halfWidth because it will be in visible viewport and it has extra
+      // buffer so that bigger labels can be visible when they are extremely at
+      // right or left.
+      if (offset.dx > -halfWidth && offset.dx < halfWidth * 3) {
         final textPainter = TextPainter(
           text: textSpan,
           textDirection: TextDirection.ltr,
         );
-
-        textPainter.layout(minWidth: 0, maxWidth: size.width);
-        textPainter.paint(
-          canvas,
-          _labels[i].offset,
-        );
+        textPainter.layout(minWidth: 0, maxWidth: halfWidth * 2);
+        textPainter.paint(canvas, offset);
       }
     }
   }
 
   void _addLabel(Canvas canvas, int i, Size size) {
-    canvas.drawLine(
-        Offset(
-            _labelPadding + dragOffset.dx - totalBackDistance.dx, size.height),
-        Offset(_labelPadding + dragOffset.dx - totalBackDistance.dx,
-            size.height + durationLinesHeight),
-        _durationLinePaint);
-    _labels.add(
-      Label(
-        content: showHourInDuration
-            ? Duration(seconds: i).toHHMMSS()
-            : Duration(seconds: i).inSeconds.toMMSS(),
-        offset: Offset(
-            _labelPadding +
-                dragOffset.dx -
-                totalBackDistance.dx -
-                durationTextPadding,
-            size.height + labelSpacing),
-      ),
-    );
+    final labelDuration = Duration(seconds: i);
+    final durationLineDx = _labelPadding + dragOffset.dx - totalBackDistance.dx;
+    final height = size.height;
+
+    if (labelDuration <
+        Duration(
+            seconds: currentlyRecordedDuration.inSeconds + durationBuffer)) {
```

Comment:
format!
Also, it may be better to extract the `Duration` object in a separate variable and then use that variable in the condition.